velox/functions/sparksql/aggregates/DecimalSumAggregate.h

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include "velox/exec/SimpleAggregateAdapter.h"
+#include "velox/type/DecimalUtil.h"
+
+namespace facebook::velox::functions::aggregate::sparksql {
+
+/// @tparam TInputType The raw input data type.
+/// @tparam TSumType The type of sum in the output of partial aggregation or the
+/// final output type of final aggregation.
+template <typename TInputType, typename TSumType>
+class DecimalSumAggregate {
+ public:
+  using InputType = Row<TInputType>;
+
+  using IntermediateType =
+      Row</*sum*/ TSumType,
+          /*isEmpty*/ bool>;
+
+  using OutputType = TSumType;
+
+  // Spark's decimal sum doesn't have the concept of a null group, each group is
+  // initialized with an initial value, where sum = 0 and isEmpty = true.
+  // Therefore, to maintain consistency, we need to use the parameter
+  // nonNullGroup in writeIntermediateResult to output a null group as sum =
+  // 0, isEmpty = true. nonNullGroup is only available when default-null
+  // behavior is disabled.
+  static constexpr bool default_null_behavior_ = false;
+
+  static bool toIntermediate(
+      exec::out_type<Row<TSumType, bool>>& out,
+      exec::optional_arg_type<TInputType> in) {
+    if (in.has_value()) {
+      out.copy_from(std::make_tuple(static_cast<TSumType>(in.value()), false));
+      return true;
+    }
+    return false;
+  }
+
+  // This struct stores the sum of input values, overflow during accumulation,
+  // and a bool value isEmpty used to indicate whether all inputs are null. The
+  // initial value of sum is 0. We need to keep sum unchanged if the input is
+  // null, as sum function ignores null input. If the isEmpty is true, then it
+  // means there were no values to begin with or all the values were null, so
+  // the result will be null. If the isEmpty is false, then if sum is nullopt
+  // that means an overflow has happened, it returns null.
+  struct AccumulatorType {
+    std::optional<int128_t> sum_{0};
+    int64_t overflow_{0};
+    bool isEmpty_{true};
+
+    AccumulatorType() = delete;
```

Comment:
this is not needed, is it?

Reply:
This is not required since another constructor with HashStringAllocator* argument is defined. But I think deleting the default constructor explicitly is still helpful to clarify the purpose of the code.